src/metabase/models/model_index.clj

```diff
@@ -0,0 +1,134 @@
+(ns metabase.models.model-index
+  (:require
+   [clojure.string :as str]
+   [metabase.db.connection :as mdb.connection]
+   [metabase.models.card :refer [Card]]
+   [metabase.models.interface :as mi]
+   [metabase.query-processor :as qp]
+   [metabase.sync.schedules :as sync.schedules]
+   [metabase.util.cron :as u.cron]
+   [metabase.util.i18n :refer [trs]]
+   [metabase.util.log :as log]
+   [toucan.models :as models]
+   [toucan2.core :as t2]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; model lifecycle ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(models/defmodel ModelIndex :model_index)
+(models/defmodel ModelIndexValue :model_index_value)
+
+(derive ModelIndex :hook/created-at-timestamped?)
+
+(t2/deftransforms ModelIndex
+  {:pk_ref    mi/transform-field-ref
+   :value_ref mi/transform-field-ref})
+
+(t2/define-before-delete ModelIndex
+  [model-index]
+  (let [remove-refresh-job (requiring-resolve 'metabase.task.index-values/remove-indexing-job)]
+    (remove-refresh-job model-index)))
+
+;;;; indexing functions
+
+(defn- fetch-values
+  [model-index]
+  (let [model (t2/select-one Card :id (:model_id model-index))]
+    (try [nil (->> (qp/process-query
+                    {:database (:database_id model)
+                     :type     :query
+                     :query    {:source-table (format "card__%d" (:id model))
+                                :fields       [(:pk_ref model-index) (:value_ref model-index)]
+                                :order-by     [[:desc (:value_ref model-index)]]
+                                :limit        5001
+                                :breakout   [(:pk_ref model-index) (:value_ref model-index)]}})
+                   :data :rows)]
+         (catch Exception e
+           (log/warn (trs "Error fetching indexed values for model {0}" (:id model)) e)
+           [(ex-message e) []]))))
+
+(defmulti add-values* "Index values in a model."
+  (fn [_model-index _values] (mdb.connection/db-type)))
+
+(defmethod add-values* :h2
+  [model-index values]
+  ;; h2 just deletes and recreates
+  (t2/delete! ModelIndexValue :model_index_id (:id model-index))
+  (t2/insert! ModelIndexValue (map (fn [[id v]]
+                                     {:name           v
+                                      :model_pk       id
+                                      :model_index_id (:id model-index)
+                                      :generation     (inc (:generation model-index))})
+                                   values)))
+
+(defmethod add-values* :postgres
+  [model-index values]
+  (let [new-generation (inc (:generation model-index))]
+    ;; use upserts and delete ones with old generations
```

Comment:
It's not just speed. I'm worried about constantly deleting rows and then adding the same ones back in. The worry is that this would accumulate lots of garbage in the table requiring a vacuum schedule. The hope is that the upserts can modify the existing row. If this proves not the case (either upserts create garbage or the vacuum worry is overblown) then yes mysql/postgres could just dump and recreate like h2 does.

Reply:
Are you sure that updates don't produce garbage? Updates in MVCC might be implemented as a delete+insert. The PostgreSQL docs contain this (vague) statement:
> VACUUM reclaims storage occupied by dead tuples. In normal PostgreSQL operation, tuples that are deleted or obsoleted by an update are not physically removed from their table; they remain present until a VACUUM is done.